core/arch/arm/kernel/generic_boot.c

```diff
@@ -793,34 +806,71 @@ static void init_fdt(unsigned long phys_fdt)
 		panic();
 	}
 
+	dt_blob_addr = fdt;
+}
+
+static void update_fdt(void)
+{
+	void *fdt = get_dt_blob();
+	int ret;
+
+	if (!fdt)
+		return;
+
 	if (add_optee_dt_node(fdt))
 		panic("Failed to add OP-TEE Device Tree node");
 
 	if (config_psci(fdt))
 		panic("Failed to config PSCI");
 
-	if (config_nsmem(fdt))
+	if (mark_static_shm_as_reserved(fdt))
 		panic("Failed to config non-secure memory");
 
 	ret = fdt_pack(fdt);
 	if (ret < 0) {
 		EMSG("Failed to pack Device Tree at 0x%" PRIxPA ": error %d",
-		     phys_fdt, ret);
+		     virt_to_phys(fdt), ret);
 		panic();
 	}
+}
 
-	dt_blob_addr = fdt;
+static void discover_nsec_memory(void)
+{
+	struct core_mmu_phys_mem *mem;
+	size_t nelems;
+	void *fdt = get_dt_blob();
+
+	if (fdt) {
```

Comment:
Sorry nitpicking, but could this be moved inside `_discover_nsec_memory()` to get a single routine for dynamic shm discovery function? It would handle all cases: fdt available or not.

Reply:
Sure, of course. I will push a new patch to fix it.